Add rename fields processor

ruflin opened the pull request:
The rename processor allows to rename fields before they are indexed to standardise on names or move fields around. This becomes useful when building filebeat modules which read from json files. With the rename processor no ingest pipeline is needed to follow the naming schema. This should make some modules simpler to build. It's also useful in combination with #6024 to rename some fields according to the schema.

```
processors:
- rename:
    fields:
     - from: "a"
       to: "b"
```

Intention of rename
* Adjust fields to mapping
* Prevent conflicts like `a` and `a.b` by renaming `a` to `a.value`

Limitations
* Will not overwrite keys

ph commented on libbeat/processors/actions/rename_fields.go:
```diff
@@ -0,0 +1,69 @@
+	}
+
+	return event, nil
+}
```
In the context of parsing multiple JSON document, I think that logging the error everytime could be quite noisy. I think we should add `ignore_missing`, this would be similar to what ingest provide if the key doesn't exist in multiple processors. See the [rename processor](https://www.elastic.co/guide/en/elasticsearch/reference/master/rename-processor.html)

ph replied:
I  see we diverge from the `ignore_missing` semantic from the ingest pipeline, in their case if the flag is set to false its a fatal error. This might be OK in our usecase.

ruflin replied:
What happens in the case of a fatal error? The original event is stored or the processed one until the error happened? We could have an additional config ignore_errors.

ph replied:
@ruflin Good question, I do not have good answer, Do we have any other existing processor that implement this kind of behavior?

ph commented on libbeat/processors/actions/rename_fields.go:
```diff
@@ -0,0 +1,69 @@
+	}
+
+	// TODO: Check if some fields are not allowed to be renamed
+
```
What fields cannot be renamed?

ruflin replied:
Not sure if we should allow to rename `@timestamp` for example.

ph replied:
@ruflin are you planning to block the following two options:
```yaml
     - from: "@timestamp"
       to: "another_timestamp"
     - from: "good_timestamp"
       to: "@timstamp"
```

The last one would require time parsing?

ruflin replied:
I was initially thinking about it but now thinking we should not add any artificial limitations here and lets see what happens.

ph commented on libbeat/processors/actions/rename_fields.go:
```diff
@@ -0,0 +1,69 @@
+	Fields []FieldTuple
+}
+
+type FieldTuple struct {
```
+1 to hide it.

ruflin replied:
done

ph commented:
+1 for `ignore_missing`, but set it to false by default.

andrewkroh commented on libbeat/processors/actions/rename_fields.go:
```diff
@@ -0,0 +1,69 @@
+	for _, field := range f.Fields {
+		value, err := event.Fields.GetValue(field.From)
+		if err != nil {
+			logp.Err("Could not fetch from key: %s", field.From)
```
Every key that does not exists will be an error. That seems to verbose to me. Maybe you want to handle the error types separately -- log a debug message if the key does not exists and log an error for anything else.

ruflin replied:
I think adding the `ignore_missing` config option should solve this? There is still the issue of what if there is an "other" error. Need to check if any other errors can happen.

ruflin commented:
I changed / simplified the config a bit and would like to get some feedback. It now looks like below:

```
processors:
- rename_fields:
    fields:
     - "a.b": "c.d"
    ignore_missing: false
- rename_fields:
    fields:
     - "a.g": "e.d"
    ignore_missing: true
```

Writing tests I realised the from / to was kind of too long and not needed. Using just key / value pairs is enough and makes it very easy to specify the values.

The ignore_missing is per processor instance, so if only for a few fields it should be ignored, the processor can be set twice.

ruflin commented on libbeat/processors/actions/rename_fields_test.go:
```diff
@@ -0,0 +1,96 @@
+			error: false,
+		},
+		{
+			// This overwrites an existing field
```
@andrewkroh @ph I wonder what our expected behaviour here is. We could later add an option "prevent overwrite" or something if it becomes an issue.

ph replied:
Well, I think we should aim for the **least surprise**, what I mean, is we should fails if the target field already exist. They can always use a `drop_field` before a rename to ensure it always works?

Reference:
https://github.com/elastic/elasticsearch/blob/master/modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RenameProcessor.java#L68-L75

ph commented on libbeat/processors/actions/rename_fields.go:
```diff
@@ -0,0 +1,76 @@
+}
+
+func (f *renameFields) Run(event *beat.Event) (*beat.Event, error) {
+
```
space?

ruflin replied:
It's going to be fine as long as you don't tell @andrewkroh :-D

andrewkroh commented on libbeat/common/mapstr.go:
```diff
@@ -322,6 +322,17 @@ func walkMap(key string, data MapStr, op mapStrOperation) (interface{}, error) {
+	// Only check for full key if key contains dots
+	if len(keyParts) > 1 {
+		// Converted dotted key in object to nested object
+		// This is needed for processors to support dotted keys
```
Can you add a test case to the MapStr tests for this. Is this needed for data that has not been de-dotted (the key names contain dots)?

andrewkroh commented on libbeat/processors/actions/rename_fields.go:
```diff
@@ -0,0 +1,101 @@
+
+	// Fields cannot be overwritten. Either the target field has to be dropped first or renamed first
+	exists, err := fields.HasKey(to)
+	// TODO: Why is it an error if HasKey is false?
```
`HasKey` returns true iff (if and only if) the key exists, so you can ignore the error.

The error is just some extra information to go along with the bool. For example `HasKey` will return an error if the object has a `x.y` and you ask for `x.y.z`, because it expects `y` to be an object, not a scalar.

andrewkroh commented on libbeat/processors/actions/rename_fields.go:
```diff
@@ -0,0 +1,101 @@
+	value, err := fields.GetValue(from)
+	if err != nil {
+		// Ignore ErrKeyNotFound errors
+		if errors.Cause(err) == common.ErrKeyNotFound && f.ignoreMissing {
```
I'd swap the order here to take advantage of the short-circuit. When `ignoreMissing==false` you can avoid calling `errors.Cause`.

andrewkroh commented on libbeat/processors/actions/rename_fields.go:
```diff
@@ -0,0 +1,101 @@
+	//	return fmt.Errorf("has key error: %s", err)
+	//}
+	if exists {
+		return fmt.Errorf("Target field %s already exists, drop or rename this field first.", to)
```
Error strings should not be capitalized. (there are few of these) https://github.com/golang/go/wiki/CodeReviewComments#error-strings

ruflin commented on libbeat/processors/actions/rename_fields.go:
```diff
@@ -0,0 +1,101 @@
+	}{}
+	err := c.Unpack(&config)
+	if err != nil {
+		return nil, fmt.Errorf("fail to unpack the rename_fields configuration: %s", err)
```
changed

ruflin commented:
@ph @andrewkroh I updated the PR with more tests, added docs and implemented the config options `ignore_missing` and `fail_on_error`. I would appreciate some feedback on the config options to hear if these are enough, too much or others are needed.

ph commented on libbeat/processors/actions/rename_fields.go:
```diff
@@ -0,0 +1,99 @@
+}
+
+func (f *renameFields) renameField(from string, to string, fields common.MapStr) error {
+
```
space?

ruflin commented on libbeat/docs/processors-using.asciidoc:
```diff
@@ -513,6 +514,41 @@ section.
+[[rename-fields]]
+=== Rename fields from events
+
+The `rename_fields` processor specifies a list of fields which should be renamed. The list of fields to be
```
done

ruflin commented:
@andrewkroh @ph New version pushed

ruflin commented:
Please do not merge it yet. I did run some additional manual tests and run into issue with dotted keys in yaml. Perhaps the config needs to change again :-(

andrewkroh commented on libbeat/docs/processors-using.asciidoc:
```diff
@@ -513,6 +514,45 @@ section.
+Having a field with the name `a:2` and `a.b:4` will return an error from
+Elasticsearch during ingest time as `a` would be an integer and object at the
+same time. To prevent this the rename processor can be used to rename `a` to
+`a.value` for example.
```
I was struggling on "Having a field with the name `a:2` and `a.b:4`" so I have a suggestion for the paragraph.

Renaming fields can be useful in cases where field names cause conflicts. For example if an event has two fields `a` and `a.b` that are both assigned a scalar values (e.g. `{"a": 1, "a.b": 2}`) this will result in an error at ingest time from Elasticsearch. This is because the value of `a` cannot simultaneously be a scalar and an object. To prevent this `rename_fields` can be used to rename `a` to `a.value`.

ruflin replied:
Done, thanks for the proposal.

andrewkroh commented on libbeat/docs/processors-using.asciidoc:
```diff
@@ -513,6 +514,45 @@ section.
+=== Rename fields from events
+
+The `rename_fields` processor specifies a list of fields that should be renamed.
+The list of fields to be renamed are listed under `fields`. Each entry contains a `from: old-key` and a `to: new-key` pair. `from` is the origin and `to` the target name of the field.
```
Can you please fix the line wrapping again. Also below, the config options lines like `ignore_missing` are long too.

ruflin replied:
Done. I hoped to put this in our editor_config to have it automatically but it's not supported yet. https://github.com/editorconfig/editorconfig/wiki/Property-research:-maximum-line-length

Also pity that IntellJ does not support it based on file types: https://youtrack.jetbrains.com/issue/IDEA-89608

ph commented:
@ruflin I took another look at this PR, concerning the `from/to` combo we could use `source/target` or we could use the same naming as the [ingest processor](https://www.elastic.co/guide/en/elasticsearch/reference/master/rename-processor.html).
```
{
  "rename": {
    "field": "foo",
    "target_field": "foobar"
  }
}
```

We could also drop the `_field` and use `rename` instead of rename field.

andrewkroh commented on libbeat/processors/actions/rename_fields.go:
```diff
@@ -0,0 +1,103 @@
+}
+
+func newRenameFields(c *common.Config) (processors.Processor, error) {
+	config := struct {
```
There's some duplication in this anonymous struct and the `renameFields` struct. Perhaps you could elevate this anonymous part to `type renameFieldsConfig struct` and then embed it in the `renameFields` processor.

```
type renameFields struct {
    renameFieldsConfig
}

type renameFieldsConfig struct {
    Fields        []fromTo `config:"fields"`
    IgnoreMissing bool     `config:"ignore_missing"`
    FailOnError   bool     `config:"fail_on_error"`
}
```

ruflin replied:
done

ruflin commented:
@ph +1 on dropping `_field`. I got inspired by `drop_fields` but the shorter version seems to be better.

For the ingest configs I struggle a bit with `target_field`. The reason I ended up with `from/to` is that it is pretty easy to read. `rename` -> `fields` -> `from`  x -> `to` -> `y`. source/target is a bit longer but also an option :-) Naming is hard.

ph commented:
> Now that you suggested removing _fields I think I'm in favor again of from/to. Naming is hard.

I am OK with that go ahead :)

ruflin commented on libbeat/_meta/config.reference.yml:
```diff
@@ -137,6 +137,14 @@
 #       equals:
 #           http.code: 200
 #
+# The following example renames the field a to b:
```
@dedemorton The processor part in the reference file becomes too extensive in my opinion. We should find a way to link here to the docs instead. WDYT?

dedemorton replied:
@ruflin I agree with you. We need to be careful about linking to docs, tho. We need to make sure the links get validated. I can investigate. We probably shouldn't hold up the merge of this PR for that, tho.

ruflin commented:
@ph @andrewkroh PR updated and rebased again on master.

dedemorton commented on libbeat/docs/processors-using.asciidoc:
```diff
@@ -531,6 +532,52 @@ section.
+[[rename-fields]]
+=== Rename fields from events
+
+The `rename` processor specifies a list of fields that should be renamed. The
```
Would be better to say, "...specifies a list of fields to rename."

ruflin replied:
changed

dedemorton commented on libbeat/docs/processors-using.asciidoc:
```diff
@@ -531,6 +532,52 @@ section.
+=== Rename fields from events
+
+The `rename` processor specifies a list of fields that should be renamed. The
+list of fields to be renamed are listed under `fields`. Each entry contains a
```
"list of fields to be renamed are listed under `fields`" seems a little redundant.

ruflin replied:
changed

dedemorton commented on libbeat/docs/processors-using.asciidoc:
```diff
@@ -531,6 +532,52 @@ section.
+`from: old-key` and a `to: new-key` pair. `from` is the origin and `to` the
+target name of the field.
+
+The renaming of fields can be useful in case two fields potentially conflict.
```
Delete this sentence? The sentence that follows seems to say the same thing

ruflin replied:
removed

dedemorton commented on libbeat/docs/processors-using.asciidoc:
```diff
@@ -531,6 +532,52 @@ section.
+
+Renaming fields can be useful in cases where field names cause conflicts. For
+example if an event has two fields, `a` and `a.b`, that are both assigned a
+scalar values (e.g. `{"a": 1, "a.b": 2}`) this will result in an error at ingest
```
"a scalar values is wrong". should be either "that are both assigned scaler values"

ruflin replied:
fixed

dedemorton commented on libbeat/docs/processors-using.asciidoc:
```diff
@@ -531,6 +532,52 @@ section.
+Renaming fields can be useful in cases where field names cause conflicts. For
+example if an event has two fields, `a` and `a.b`, that are both assigned a
+scalar values (e.g. `{"a": 1, "a.b": 2}`) this will result in an error at ingest
+time from Elasticsearch. This is because the value of a cannot simultaneously be
```
wording here is a little confusing. I think you want to say: "...this will result in an Elasticsearch error at ingest time"

ruflin replied:
change

dedemorton commented on libbeat/docs/processors-using.asciidoc:
```diff
@@ -531,6 +532,52 @@ section.
+Renaming fields can be useful in cases where field names cause conflicts. For
+example if an event has two fields, `a` and `a.b`, that are both assigned a
+scalar values (e.g. `{"a": 1, "a.b": 2}`) this will result in an error at ingest
+time from Elasticsearch. This is because the value of a cannot simultaneously be
```
Shouldn't `a` be wrapped in back ticks here? In general, I'd avoid using "a" as a field name in an example because it's so easy to confuse it with a part of speech (article) rather than a field name.

ruflin replied:
added ticks and replaced by `c`.

dedemorton commented on libbeat/docs/processors-using.asciidoc:
```diff
@@ -531,6 +532,52 @@ section.
+example if an event has two fields, `a` and `a.b`, that are both assigned a
+scalar values (e.g. `{"a": 1, "a.b": 2}`) this will result in an error at ingest
+time from Elasticsearch. This is because the value of a cannot simultaneously be
+a scalar and an object. To prevent this rename_fields can be used to rename a to
```
same comment here. Use back ticks to wrap field names, especially if they can be confused with articles.

dedemorton commented on libbeat/docs/processors-using.asciidoc:
```diff
@@ -531,6 +532,52 @@ section.
+a.value.
+
+Rename fields cannot be used to overwrite fields. To overwrite fields either
+first rename the target field or use the drop_fields processor to drop the field
```
wrap processor names in back ticks: use the `drop_fields` processor

ruflin replied:
done

ruflin commented:
Updated version pushed.
